swri_roscpp/include/swri_roscpp/topic_service_client.h

```diff
@@ -0,0 +1,173 @@
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Southwest Research Institute® (SwRI®) nor the
+//       names of its contributors may be used to endorse or promote products
+//       derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL SOUTHWEST RESEARCH INSTITUTE BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// *****************************************************************************
+#ifndef SWRI_ROSCPP_TOPIC_SERVICE_CLIENT_H_
+#define SWRI_ROSCPP_TOPIC_SERVICE_CLIENT_H_
+
+#include <ros/node_handle.h>
+
+#include <boost/thread/mutex.hpp>
+
+#include <map>
+
+namespace swri
+{
+
+template<class MReq, class MRes>
+class TopicServiceClientRaw
+{
+ private:
+  typedef
+  boost::mutex request_lock_;
+  ros::Subscriber response_sub_;
+  ros::Publisher request_pub_;
+  boost::shared_ptr<MRes> response_;
+
+  ros::Duration timeout_;
+  std::string name_;
+  std::string service_name_;
+
+  int sequence_;
+
+ public:
+  TopicServiceClientRaw() : sequence_(0), timeout_(ros::Duration(4.0))
+  {
+
+  }
+
+  void initialize(ros::NodeHandle &nh,
+                const std::string &service,
+                const std::string &client_name = "")
+  {
+    name_ = client_name.length() ? client_name : ros::this_node::getName();
+    std::string rservice = nh.resolveName(service);
+    service_name_ = rservice;
+    response_sub_ = nh.subscribe(rservice + "/response", 10, &TopicServiceClientRaw<MReq, MRes>::response_callback, this);
+    request_pub_ = nh.advertise<MReq>(rservice + "/request", 10);
+  }
+
+  bool call(MReq& request, MRes& response)
+  {
+    boost::mutex::scoped_lock scoped_lock(request_lock_);
+
+    // block for response
+    request.srv_header.stamp = ros::Time::now();
+    request.srv_header.sequence = sequence_;
+    request.srv_header.sender = name_;
+
+    // Wait until we get a subscriber and publisher
+    while (request_pub_.getNumSubscribers() == 0 && response_sub_.getNumPublishers() == 0)
```

Comment:
This will only wait until you get a subscriber *or* a publisher, since it will return false as soon as either one of them is >0.  Is that what you intended?

Reply:
Good catch, it was not. Fixing